google/cloud/firestore_bundle/bundle.py

```diff
@@ -0,0 +1,439 @@
+# Copyright 2021 Google LLC All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Classes for representing bundles for the Google Cloud Firestore API."""
+
+import datetime
+import json
+
+from google.cloud.firestore_bundle.types.bundle import (
+    BundledDocumentMetadata,
+    BundledQuery,
+    BundleElement,
+    BundleMetadata,
+    NamedQuery,
+)
+from google.cloud._helpers import _datetime_to_pb_timestamp, UTC  # type: ignore
+from google.cloud.firestore_bundle._helpers import limit_type_of_query
+from google.cloud.firestore_v1.async_query import AsyncQuery
+from google.cloud.firestore_v1.base_client import BaseClient
+from google.cloud.firestore_v1.base_document import DocumentSnapshot
+from google.cloud.firestore_v1.base_query import BaseQuery
+from google.cloud.firestore_v1.document import DocumentReference
+from google.cloud.firestore_v1.query import Query
+from google.cloud.firestore_v1 import _helpers
+from google.protobuf.timestamp_pb2 import Timestamp  # type: ignore
+from typing import (
+    Dict,
+    Generator,
+    List,
+    Optional,
+    Union,
+)
+
+
+class FirestoreBundle:
+    """A group of serialized documents and queries, suitable for
+    longterm storage.
```

Comment:
Note that because of how resume tokens work in Firestore, it is cheaper to update a bundle more often. If a bundle is less than 30 minutes old, queries can be resumed cheaply on consuming clients.

"Longterm storage" is still a thing though, but we think that customers will frequently want to update their bundles.

Reply:
It makes sense that documents updated intermittently, but read very frequently, would be updated regularly. I'd previously imagined this as mostly being used for lookup tables, and the like.

At any rate, I've expanded on the docstring's opening sentence.